Add API and UI to remove requirements installed by Conda

mvdbeek opened the pull request:
This is currently limited to the Conda dependency resolver, but other
resolver would only need to implement the resolver.uninstall method and
set can_uninstall_dependencies to True.

To use this endpoint, DELETE to
`<galaxy_url>/tools/<tool_id>/dependencies`.

In this PR I have also added the possibility to POST to <galaxy_url>/tools/<tool_id>/dependencies, which will install the dependencies, in favor of the older `<galaxy_url>/tools/<tool_id>/install_dependencies` endpoint which we will deprecate (when ?).

If this is accepted, I will enhance the new manage tool dependencies page so that we can remove and reinstall dependencies from there.

mvdbeek committed:
Fix typo in install_dependencies docstring

mvdbeek committed:
Add API to remove requirements installed by Conda

This is currently limited to the Conda dependency resolver, but other
resolver would only need to implement the resolver.uninstall method and
set can_uninstall_dependencies to True.

To use this endpoint, POST to
<galaxy_url>/tools/<tool_id>/uninstall_dependencies'.

mvdbeek committed:
Return last non-zero returncode

jmchilton commented:
@mvdbeek Are you willing to work on these endpoint URLs a bit?

I'd like to maintain

```
POST /api/tools/{tool_id}/install_dependencies
```

for backward compat. but I guess going forward I'd like to see the following for install and uninstall:

```
POST /api/tools/{tool_id}/dependencies
DELETE /api/tools/{tool_id}/dependencies
```

It feels more RESTful to me.

mvdbeek commented:
Oh, sure, that makes sense!

mvdbeek committed:
Change endpoint to `api/tools/<tool_id>/dependencies`

A POST will trigger installation and DELETE will trigger uninstalling
the dependencies (Thx for the suggestion @jmchilton).

jmchilton commented:
Did you mean to add #3729 into this PR?

mvdbeek commented:
damnit, too much switching branches :(

mvdbeek commented:
Alright, this should be OK now.

martenson reviewed (approved):
looks great

martenson commented:
should we add `POST /api/tools/{tool_id}/install_dependencies` to https://github.com/galaxyproject/galaxy/issues/3280 ?

jmchilton commented:
@martenson I'd say -0 on deprecation - the amount of technical debt we reduce is not worth the effort and the cost of not maintaining backward compat.

Update: I mean I'm fine if we all agree it is deprecated - I just don't think it is worth removing per se.

martenson commented:
@jmchilton not worth removing - worth announcing though?

mvdbeek committed:
Fix uninstall when not all dependencies are resolved by all-in-one resolver

mvdbeek committed:
Add UI for Conda dependency uninstall

mvdbeek commented:
I have fixed a small problem in 6902e62 and I have added a button to actually uninstall the conda env in the UI in 95106ea.

martenson commented:
Thanks @mvdbeek !

nsoranzo commented on lib/galaxy/tools/deps/resolvers/conda.py:
```diff
@@ -116,6 +117,25 @@ def get_option(name):
+        final_return_code = 0
+        for env, return_code in zip(environments, return_codes):
+            if return_code == 0:
+                log.debug("Conda environment '%s' sucessfully removed." % env)
```
s/sucessfully/successfully/

nsoranzo commented on lib/galaxy/tools/deps/resolvers/conda.py:
```diff
@@ -116,6 +117,25 @@ def get_option(name):
 
+    def uninstall(self, requirements):
+        """Uninstall requirements installed by install_all or multiple install statements."""
+        all_resolved = [r for r in self.resolve_all(requirements) if r.dependency_type]
```
This will break when `self.resolve_all()` returns `False` in line 163.
